signer/universal/universal_test.go

```diff
@@ -38,5 +287,345 @@ func TestNewSigner(t *testing.T) {
+				}
+				cert, err := helpers.ParseCertificatePEM(certBytes)
+				if err != nil {
+					t.Fatal("Fail to parse returned certificate:", err)
+				}
+				ku, _, _ := profile.Usages()
+				if cert.KeyUsage != ku {
+					t.Fatalf("Key usage was incorrect expected %+v, got %+v", ku, cert.KeyUsage)
+				}
+			}
+		}
+	}
+}
+
+func newConfig(t *testing.T, configBytes []byte) *config.Config {
+	conf, err := config.LoadConfig([]byte(configBytes))
+	if err != nil {
+		t.Fatal("config loading error:", err)
+	}
+	if !conf.Valid() {
+		t.Fatal("config is not valid")
+	}
+	return conf
+}
+
+func newTestUniversalSigner(t *testing.T, policy *config.Signing) signer.Signer {
+	h := map[string]string{
+		"key-file":  testCaKeyFile,
+		"cert-file": testCaFile,
+	}
+
+	r := &Root{
+		Config:      h,
+		ForceRemote: false,
+	}
+
+	s, err := NewSigner(*r, policy)
+	if err != nil {
+		t.Fatal("fail to init universal signer:", err)
+	}
+
+	return s
+}
+
+func newTestSignHandler(t *testing.T, s signer.Signer) (h http.Handler) {
+	h, err := NewSignHandlerFromSigner(s)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return
+}
+
+func newTestInfoHandler(t *testing.T, s signer.Signer) (h http.Handler) {
+	h, err := apiinfo.NewHandler(s)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return
+}
+
+func newTestSignServer(t *testing.T, s signer.Signer) *httptest.Server {
+	mux := http.NewServeMux()
+	mux.Handle("/api/v1/cfssl/sign", newTestSignHandler(t, s))
+	ts := httptest.NewUnstartedServer(mux)
+	ts.Start()
+	t.Log(ts.URL)
+	return ts
+}
+
+func newTestInfoServer(t *testing.T, s signer.Signer) *httptest.Server {
+	mux := http.NewServeMux()
+	mux.Handle("/api/v1/cfssl/info", newTestInfoHandler(t, s))
+	ts := httptest.NewUnstartedServer(mux)
+	ts.Start()
+	t.Log(ts.URL)
+	return ts
+}
+
+func closeTestServer(t *testing.T, ts *httptest.Server) {
+	t.Log("Finalizing test server.")
+	ts.Close()
+}
+
+// NewSignHandlerFromSigner generates a new SignHandler directly from
+// an existing signer.
+func NewSignHandlerFromSigner(s signer.Signer) (h http.Handler, err error) {
+	policy := s.Policy()
+	if policy == nil {
+		err = errors.New(errors.PolicyError, errors.InvalidPolicy)
+		return
+	}
+
+	// Sign will only respond for profiles that have no auth provider.
+	// So if all of the profiles require authentication, we return an error.
+	haveUnauth := (policy.Default.Provider == nil)
+	for _, profile := range policy.Profiles {
+		if !haveUnauth {
+			break
+		}
+		haveUnauth = (profile.Provider == nil)
+	}
+
+	if !haveUnauth {
+		err = errors.New(errors.PolicyError, errors.InvalidPolicy)
+		return
+	}
+
+	return &api.HTTPHandler{
+		Handler: &SignHandler{
+			signer: s,
+		},
+		Methods: []string{"POST"},
+	}, nil
+}
+
+// A SignHandler accepts requests with a hostname and certficate
+// parameter (which should be PEM-encoded) and returns a new signed
+// certificate. It includes upstream servers indexed by their
+// profile name.
+type SignHandler struct {
```

Comment:
Why create a new SignHandler type in the test rather than use the one from the api/sign package?